Added API functions for ASI API implemenation. Including import and e…

PavelStishenko opened the pull request:
This PR is a rebased version of the PR #1093 .
It includes all the work done to enable ASI API implementation. Mainly C API extension to include export and import of the Hamiltonian, Overlap, and Density matrices.

Commit:
Added API functions for ASI API implemenation. Including import and export of Hamitlonian, Overlap and Density Matrices.

Commit:
Added API functions for ASI API implemenation. Including import and export of Hamitlonian, Overlap and Density Matrices.

bhourahine commented on CMakeLists.txt:
```diff
@@ -183,7 +183,7 @@ string(REGEX MATCH "(^|;)[Ss]ubmodule(^|;)" INCLUDE_INDIRECT_DEPS "${HYBRID_CONF
 if(WITH_MPI)
   set(MPIFX_GIT_REPOSITORY "https://github.com/dftbplus/mpifx.git")
-  set(MPIFX_GIT_TAG "da51073aa87831a91ae756a9773e39ea000d1c3a")  # do not change manually!
+  set(MPIFX_GIT_TAG "0641eead6ff9bf04c7ca2ec676a281780e2029e2")  # do not change manually!
```
This is behind the main branch dependency (from before the rebase time).

PavelStishenko replied:
Fixed!

bhourahine commented on src/dftbp/api/mm/apicallbackc.f90:
```diff
@@ -0,0 +1,79 @@
+!--------------------------------------------------------------------------------------------------!
+!  DFTB+: general package for performing fast atomistic simulations                                !
+!  Copyright (C) 2006 - 2022  DFTB+ developers group                                               !
```
Can you update copyright data on all files to 2023

PavelStishenko replied:
Fixed!

bhourahine commented on test/src/dftbp/api/mm/testers/test_mpiexpdmhs.c:
```diff
@@ -0,0 +1,150 @@
+  for (int i = 0; i < BASIS_SIZE; ++i)
+    for (int j = 0; j < BASIS_SIZE; ++j) {
+      hs_type d = dm[i][j];
+      // if (i != j)
```
Are these two lines redundant comments, or of use for some past testing?

bhourahine commented on test/src/dftbp/api/mm/testcases/mpiexpdmhs/dftb_in.O2.hsd:
```diff
@@ -0,0 +1,55 @@
+  Blacs {
+    BlockSize = 4
+  }
+  Groups = 2
```
Might be a good idea to have another test case where the groups are used for k-parallelism instead of spin. The Si2 example you've added could form the basis of this.

bhourahine commented on test/src/dftbp/api/mm/testcases/mpiexpdmhs/dftb_in.O2.hsd:
```diff
@@ -0,0 +1,55 @@
+}
+
+ParserOptions = {
+  ParserVersion = 9
```
```suggestion
  ParserVersion = 12
```

PavelStishenko replied:
Fixed

bhourahine commented on test/src/dftbp/api/mm/testcases/imp_ham/dftb_in.h2o.hsd:
```diff
@@ -0,0 +1,25 @@
+   WriteResultsTag = Yes 
+} 
+ParserOptions { 
+   IgnoreUnprocessedNodes = Yes  
```
This is missing a parser version, so will eventually break if the HSD input changes in the future

PavelStishenko replied:
Fixed

bhourahine reviewed (commented):
See a list of mostly cosmetic changes (and a couple of suggestions). Hopefully we can get the final review done soon.

PavelStishenko commented on src/dftbp/api/mm/apicallbackc.f90:
```diff
@@ -0,0 +1,79 @@
+  private
+  public :: dmhs_callback_c_wrapper, TCAuxWrapper
+
+  type :: TCAuxWrapper
```
Added!

PavelStishenko commented on src/dftbp/dftbplus/initprogram.F90:
```diff
@@ -1109,6 +1110,10 @@ module dftbp_dftbplus_initprogram
     !> of the atoms via the API is forbidden.
     logical :: atomOrderMatters = .false.
+    
+    !> This object incapsulates subroutines and variables that are used for registering and 
```
Fixed

PavelStishenko commented on src/dftbp/dftbplus/apicallback.F90:
```diff
@@ -0,0 +1,213 @@
+  end interface
+
+  
+  !> This type incapsulates registering and invocation of callbacks for export of the density,
```
Fixed

PavelStishenko commented on test/src/dftbp/api/mm/testcases/expdmhs/dftb_in.Si2.hsd:
```diff
@@ -0,0 +1,51 @@
+}
+
+ParserOptions = {
+  ParserVersion = 6
```
Fixed

PavelStishenko commented:
I believe that all suggestions are accepted or fixed.

bhourahine committed:
Comments and hybrid H fix

Comment up some routines and variables
Fix H with the range separated hamiltonian
Fix a few comments (copyright year and minor typos)

bhourahine committed:
Comments and hybrid H fix

Comment up some routines and variables
Fix H with the range separated hamiltonian
Fix a few comments (copyright year and minor typos)

bhourahine commented:
@PavelStishenko I've made a PR into your branch (instead of listing lots of minor changes). As mentioned in that PR, there are then  a few things to discuss on top of this, mostly to lock down users from performing broken calculations if the matrices are imported.

PavelStishenko committed:
Merge pull request #1 from bhourahine/BHcomments1

Comments and hybrid H fix

bhourahine commented:
OK, just to keep them, listing the comments from PR#1 into this:
- the 'invokeX' method isn't too informative as a name for the API calls, is it an exchange, import or export of quantities?
- Please export double precision values for the expdmhs autotest.tag data.
- Cases that will fail to work correctly will need to be locked down (mostly property evaluations using the sparse S, as this isn't constructed, but this could be constructed given the k-dependent S, but we'll need to know the spatial cut-off of matrix elements).
- If the callback gets registered, various invoked modes of the code will need to be locked out of being used (mostly excited state methods).

bhourahine commented:
@PavelStishenko Are there any security issues for the current structure of leaving an open set of calls in the binary? There don't seem to be any obvious buffer overrun possibilities. I guess we can at least add a flag inside the DFTB+ input to disable the callback register and only build it in if the api is enabled.

bhourahine committed:
Check for API inclusion when adding callback

PavelStishenko commented:
> @PavelStishenko Are there any security issues for the current structure of leaving an open set of calls in the binary? There don't seem to be any obvious buffer overrun possibilities. I guess we can at least add a flag inside the DFTB+ input to disable the callback register and only build it in if the api is enabled.

If I understand your concerns  correctly, then for sure  there are no security issues due to API functions. This kind of API can be used only when the DFTB+ library is already in the address space of the process that is able to do API calls.  In that case the process is capable to whatever in wants with the library. It is by design. Of course, if there is a bug in the process that loads and uses the DFTB+ library, then it can corrupt anything in its address space, including DFTB+ data, but I don't think that it is a problem of the library.

Perhaps it is possible to compile API support conditionally, but I don't see the problem that is supposed to be solved in that way. It will require more efforts to support and test such code. And definitely will take a time to implement.

bhourahine commented:
The API can be conditionally compiled, but not the ASI parts. I think it is also safer if the user input can also turn the feature on and off for enabled binaries (this is what i-PI does).
I'll make a PR into your branch with the changes for this (rebased on the current main code) in a few minutes.

bhourahine committed:
ASI now included in API conditional compilation

Also add a keyword in the main code to turn ASI bindings on/off
at run-time (and adjusted test cases to match).

bhourahine committed:
Dynamic allocation for expdmhs

Sort out triangles in output by symmetrizing.

Weighted integral over k for electrons and band structure.

Molecular test case added

PavelStishenko committed:
Merge pull request #3 from bhourahine/expdmhsDynamic

Dynamic allocation for expdmhs

Commit:
Merge remote-tracking branch 'ben/api-H-import-rebase2' into bhourahine-api-H-import-rebase2

Commit:
Merge branch 'bhourahine-api-H-import-rebase2' into api-H-import-rebase

PavelStishenko committed:
Merge branch 'dftbplus:main' into api-H-import-rebase

PavelStishenko committed:
Merge branch 'main' into api-H-import-rebase

Commit:
Fix optional callback

Commit:
Fix  nrOfSpin

bhourahine commented:
Functionality already merged in #1485
